lib/utils.js

```diff
@@ -39,6 +39,21 @@ exports.pick = function (obj, arr) {
   return res;
 };
 
+// just like emitter.once(event, listener), but doesn't
+// add additional listeners if any listeners with the
+// key are already attached
+exports.oncePerKey = function oncePerKey(emitter, event, key, listener) {
+  var listeners = emitter.listeners(event).filter(function (listener) {
+    return listener._key === key;
+  });
+  if (listeners.length) {
+    return;
+  }
+  emitter.once(event, listener);
+  listeners = emitter.listeners(event);
+  listeners[listeners.length - 1]._key = key;
+};
+
```

Comment:
Yes the remove - add strategy would be more resilient in case someone decides to make the proper functioning of the cleanup functions depend on the value of the closure variables.

Does the replication behave correctly if the source or destination fires 'destroy' while backOff is waiting?